### ISSUE TITLE
SDL2 version: screenshots size is always 1600x1200 regardless of the display settings

### ISSUE THREAD

**Issue by Zodomaniac** [chocolate-doom/chocolate-doom#920]
Fixing aspect ratio or not, fullscreen or not, with any window size, the PNG screenshots are always 1600x1200.

**Comment by linguica** [comment]
This is intentional since 1600x1200 is the smallest resolution that perfectly captures the original image with aspect ratio correction applied.

**Comment by jmtd** [comment]
We intended this for the combination of PNG screenshots and aspect_ratio_correct. I think we planned for the non aspect ratio correct behaviour to match your chosen resolution but there are some sdl-image bugs that prevented us doing that yet.

The file size for the 1600x1200 PNGs is pretty good (they're smaller than the equivalent 320x200 PCX files)

**Comment by AXDOOMER** [comment]
This is not a bug. Disable the "Save screenshot in PNG format" box in Chocolate-Setup to revert to the original game's behavior if it bothers you. This new feature was made to make it easier to share pictures of vanilla Doom, since you don't have to convert the format and fix the aspect ratio using another tool.

**Comment by chungy** [comment]
Closing this as it's not a bug, but a feature ;) If you want 1:1 pixel correspondence, use the PCX format.

**Comment by Zodomaniac** [comment]
@AXDOOMER
>This new feature was made to make it easier to share pictures of vanilla Doom
I get it, but this feature is kinda incomplete, it doesn't work as expected, it shouldn't fix aspect ratio if 'Fix aspect ratio' feature is disabled, but preserve the game screen resolution instead (e.g. if it runs full-screen on 1920x1080 monitor, the game aspect ration is 16:10, i.e. the game screen size is 1728x1080, and this should also be the screenshot size).

**Comment by AXDOOMER** [comment]
No, it shouldn't preserve the game's screen resolution. This would be even worse and anti-vanilla. 

The way the current feature works was already discussed and even a purist like me agrees that it's a great feature (I was against it at first). It's the expected behavior and this feature (PNG screenshots) was added for this purpose (aspect-ratio corrected screenshots). If you don't like it, you can disable the aspect-ratio correction and resize the picture to the resolution that you like using an external tool.

**Comment by jmtd** [comment]
@AXDOOMER, honestly we never intended to force a 4:3 screenshot if the player had not turned on aspect ratio correction.

(I always play with aspect ratio correction but I guess others do not?)

@Zodomaniac 
> I hope you will eventually manage to fix it this way 👍

Basically we're tracking an SDL2 bug and when that's fixed we can revisit this I think.

**Comment by chungy** [comment]
Somewhat sad about this, the PNG specification has a [pHYs chunk](http://libpng.org/pub/png/spec/1.2/PNG-Chunks.html#C.pHYs) that would allow the file to store 320x200 images while specifying that the display should be 4:3. Unfortunately very few applications ever take this into account -- no web browsers, not Photoshop or GIMP, and so forth. Which goes into the hack of working around by storing 1600x1200 images just so things look correct.

**Comment by AXDOOMER** [comment]
> we never intended to force a 4:3 screenshot if the player had not turned on aspect ratio correction.

I never said otherwise.

**Comment by jmtd** [comment]
@chungy yes it is a shame. On the plus side the duplicate pixels compress very efficiently so there's almost no file size difference.

**Comment by fragglet** [comment]
> I get it, but this feature doesn't work as expected, it shouldn't fix aspect ratio if 'Fix aspect ratio' setting is disabled, but preserve the game screen resolution instead (e.g. if it runs full-screen on 1920x1080 monitor, the game aspect ratio is 16:10, i.e. the game screen size is 1728x1080, and this should also be the screenshot size).

I understand why that might seem like a behavior you'd expect (it's what other source ports do), but I don't think it's appropriate for Chocolate Doom. What we really want is the ability to dump the 320x200 screen (like how vanilla does). In the case of 1600x1200, it's a convenient way of doing this in a lossless way that also corrects the aspect ratio.

What you're describing would be a lossy screenshot dump, since for most screen resolutions, non-integer scaling in at least one dimension will mean that the original screen is distorted (contains blurry/interleaved pixels). This is good enough during gameplay but screenshots ought to be precise and lossless. It also fits with the new video mode behavior we're adopting for 3.0, where your physical screen resolution is an irrelevant detail that you shouldn't really worry about (we've removed the ability to configure it in the setup tool).

I'd argue that the current behavior is the best behavior for Chocolate Doom. If you *really* want to get a screenshot at your screen's physical resolution, you can probably use something like ImageMagick to scale your screenshots for you automatically.

**Comment by Zodomaniac** [comment]
@fragglet if the goal is to avoid non-integer screenshot scaling, then why isn't it possible to store e.g. 1600x1000 screens from windowed mode which is integer scaling?

**Comment by jmtd** [comment]
I thought this was already covered in the comments: we tried that but hit a fairly serious SDL bug

**Comment by chungy** [comment]
I'd argue to just save to the PCX format if you want completely resolution- and aspect-uncorrected images directly from Chocolate Doom. At the same time, it should be no big deal to scale the PNGs to whatever size you want.

**Comment by fragglet** [comment]
> why isn't it possible to store e.g. 1600x1000 screens from windowed mode (which is integer scaling)

So how would this work? If the screen is an integer factor then you get that screen resolution, otherwise it's 1600x1200? To me that sounds needlessly confusing / complicated behavior. At least "always 320x200 or 1600x1200" is straightforward enough to understand.

I'd like to flip the original question around and understand the motivations behind this feature request. Is there some particular reason why you need screenshots to match your screen resolution? What are you trying to do with your screenshots that motivates this? If there's a particular use case that we're missing with the current functionality we'd certainly like to hear it.

**Comment by Zodomaniac** [comment]
@fragglet 
>Is there some particular reason why you need screenshots to match your screen resolution?

Yes, there is: I'd like to see the same picture with same proportions both in game and on screenshots :)

**Comment by jmtd** [comment]
I cannot see any point in a non corrected screenshot output at other than 320x200.

**Comment by Zodomaniac** [comment]
@jmtd OK, let it be unscaled)

**Comment by fragglet** [comment]
> Yes, there is: I'd like to see the same picture with same proportions both in game and on screenshots :)

I understand what you're asking for, what I'm trying to understand is why. If you don't have a compelling use case ("I can't do X without this because ...") then probably the current behavior is fine.

One more thing - to address @chungy's comment:

> Somewhat sad about this, the PNG specification has a pHYs chunk that would allow the file to store 320x200 images while specifying that the display should be 4:3

In theory this sounds like the perfect solution, but if most programs don't support this then probably it's the wrong approach. In particular you mention that most web browsers do not support it, that sounds like a minimum requirement. If people can't share Chocolate Doom screenshots online and have them appear at the correct aspect ratio then it's not a good solution.

**Comment by chungy** [comment]
> In theory this sounds like the perfect solution, but if most programs don't support this then probably it's the wrong approach.

Yeah, I'm just venting frustration that only the bare minimum of the PNG spec is ever implemented ;)

**Comment by Zodomaniac** [comment]
@fragglet 
>If you don't have a compelling use case ("I can't do X without this because ...") then probably the current behavior is fine.

Isn't the fact I can't capture a screenshot with original Doom 16:10 aspect ratio when the game window/screen has one enough? Aspect ratio correction distorts the picture by squeezing it horizontally, and I play without aspect ratio correction to see the picture as game developers meant it to look. Why deprive the fans of the original proportions of the possibility to take screenshots that preserve them?

**Comment by Edward850** [comment]
@Zodomaniac 
> Aspect ratio correction distorts the picture by squeezing it horizontally, and I play without aspect ratio correction to see the picture as game developers meant it to look.
Problem with that statement; The game and art was intended to be stretched to 4:3 at all times. That's just how CRTs worked with resolution back in the day.
Required reading: https://doomwiki.org/wiki/Aspect_ratio

**Comment by chungy** [comment]
> Isn't the fact I can't capture a screenshot with original Doom 16:10 aspect ratio when the game window/screen has one enough?

This has been answered in the thread a handful of times. Use the PCX format for an uncorrected (16:10) image.

**Comment by fabiangreffrath** [comment]
> we tried that but hit a fairly serious SDL bug

I cannot remember this fairly serious SDL bug, would you mind to elaborate? Is it possible you are mixing this up with the migration to SDL_Image?

What prevents us from doing something like this?
```
--- a/src/v_video.c
+++ b/src/v_video.c
@@ -737,8 +737,8 @@ void WritePNGfile(char *filename, byte *data,
     byte *rowbuf;

     // scale up to accommodate aspect ratio correction
-    width = inwidth * 5;
-    height = inheight * 6;
+    width = inwidth * (aspect_ratio_correct ? 5 : 1);
+    height = inheight * (aspect_ratio_correct ? 6 : 1);

     handle = fopen(filename, "wb");
     if (!handle)

```

**Comment by fabiangreffrath** [comment]
>  At least "always 320x200 or 1600x1200" is straightforward enough to understand.

I must say I am rally disappointed with the way this request has been answered here!

Nobody has yet replied why the screenshot resolution and aspect-ratio is hard-wired to the choice of picture format, i.e. why it is ((PNG and 1600x1200) or (PCX and 320x200)), instead of having the sreenshot resolution and aspect ratio connected to the game's `aspect_ration_correct` setting.

**Comment by AXDOOMER** [comment]
PCX screenshots will always be 320x200, because that's vanilla behavior. 

The PNG screenshot were also intended to be always 320x200, but with the pHYs chunk indicating the correct aspect ratio. The pHYs chunk didn't work, so they decided to make it 1600x1200 only. The aspect-ratio is always corrected to be like the game was intended to be displayed, whatever your personal preference. 

Now we are there. I hope this helps. 

TL;DR: `aspect_ration_correct` has nothing to do with the PNG output resolution, it only affects your display.

**Comment by chungy** [comment]
@fabiangreffrath is specifically asking why we can't have 1600×1200 PCXes or 320×200 PNGs, dependent on aspect correction setting rather than file format. I think it's a good question, I don't have a good answer.

**Comment by AXDOOMER** [comment]
I didn't see it that way. Why would anyone want to save a 1600x1200 image in the PCX format? 

Sure, you could have 1600×1200 PCXes and 320×200 PNGs, but what would be the advantage(s)? All the pros and cons about these two specific resolutions for each format were previously debated. 

You obviously get a 320x200 PCX because that's the original screenshot format from Doom. If you get a 1600x1200 PNG, it's because it was what best answered the needs of the users who requested support for this format. 

To answer more specifically to the question: If the resolution were to depend on the aspect correction setting rather than file format, then it would break the behavior that purists expect when saving to PCX when they are playing with aspect ration correction turned on. In the case of the PNG format, then playing with the aspect ration correction turned off (saving a 320x200 picture) would defeat the reason why this format was added in the first place, so there would have been no reason to add PNG support. 

That's the best answer I can give. Personally, I don't even see the need of the PNG format since a 320x200 PCX image can be converted and rescaled easily, but this argument was debated back when the feature was implemented.

**Comment by Zodomaniac** [comment]
@AXDOOMER 
>Why would anyone want to save a 1600x1200 image in the PCX format?

Why not have always 320x200 PCX and aspect ratio correction-dependent size (1600x1200 if enabled and 320x200 else) PNG?

**Comment by AXDOOMER** [comment]
> IMO when saving a PNG the player expects to get a screenshot matching the in-game aspect ratio without using any extra tools.

I understand your perspective, but having screenshots taken in the __incorrect__ aspect ratio is wrong and is not what most players expect (except for PCX because *vanilla behavior*). The aspect ratio setting only affects what is displayed to you on your screen and it would probably bring more confusion if it did affect the PNG screenshot resolution. (PNG screenshots are meant to always show Doom like it's supposed to be displayed, so it's suitable for the Doom Wiki, etc.)

There could be more customization options added to Chocolate-Doom, but it's Chocolate-Doom philosophy not to be bloated with features that try to satisfy everyone's preferences. 

You shouldn't even be playing without the aspect ratio correction if you're not using a CRT monitor. Chocolate-Doom's "Fix aspect ratio" setting emulates a CRT monitor which took the 320x200 input and displayed it as a 4:3 aspect ratio. When disabled, it's meant to be used with CRT monitors which do the 4:3 scaling by themselves. 

To summarize, because you are playing with the wrong aspect ratio, you are expecting the PNG screenshots that you capture to be saved in the wrong aspect ratio as well. Sorry.

**Comment by fabiangreffrath** [comment]
I think he isn't wrong to expect that, it's called "**screen**shot" after all.

**Comment by AXDOOMER** [comment]
> I think he isn't wrong to expect that, it's called "screenshot" after all.

You're right that it's a **screen**shot, but if you think about it like that, then it opens the door for the same feature request to be made for PCX formated screenshots. 

If you make this argument in favor of this feature for the PNG **screen**shots, then it necessarily has to apply to the PCX **screen**shots. It doesn't logically fit into Chocolate-Doom's philosophy.

**Comment by jmtd** [comment]
> I cannot remember this fairly serious SDL bug, would you mind to elaborate? Is it possible you are mixing this up with the migration to SDL_Image?

I didn't think I was mixing it up, but that was exactly the problem. You brought in sdl-image to rewrite all the screenshot code and also use the screen size/ratio for the screenshots, but the SDL_image bugs we hit (no working screenshots on OS X was one problem) meant we had to revert it.

> What prevents us from doing something like this?

I wouldn't oppose that. I don't have a strong feeling about what happens to PNG screenshots if `aspect_ratio_correct` is off. I'm quite surprised @AXDOOMER is so passionate about that case.

**Comment by AXDOOMER** [comment]
I just think that it would be very bad software design to determine if the PNG screenshot will have a 320x200 resolution or a 1600x1200 resolution on the basis of the `aspect_ratio_correct` setting, which is a display setting and shouldn't have anything to do with the resulting screenshot, that's why I have a strong opposition. I consider that this would be taking decisions on behalf of the user. The current PNG screenshot aspect ratio is how Doom was meant to be displayed, so there's nothing wrong with it, as it concords with Chocolate-Doom's goal. 

No one as complained before and the PNG screenshot feature has been working like that since 2 years. There could be a [compromise](https://github.com/chocolate-doom/chocolate-doom/pull/580#issuecomment-121818551), but I don't really agree with people who want to add new settings to satisfy everyone's requests, particularly when you can satisfy your needs in other ways.

**Comment by JNechaevsky** [comment]
The discussion comes very far, so I have to add my two cents.

Well, in my opinion, WYSIWYG logic is (will be*) quite good for screen shots, i.e. if you are playing in 1920x1080 screen resolution, you should get 1920x1080 screenshot of what you are actually see on the screen. But it is not possible because of some SDL bug, if I got all the things right.

As for PCX, **nothing** should not be changed and strictly left "as is" for Philosophy reasons.

The only thing I really want is to this case will not become any kind of 3.0 blocker. 
Also, I must admit that I'm always using screen recording software for making fullscreen screenshots, and Alt+Win combination for windowed mode.

**Comment by fabiangreffrath** [comment]
Thank you, but I don't think anyone ever seriously asked for aspect ratio corrected PCX screenshots. As far as I understood it, the entire discussion was about PNG screenshots respecting the aspect ratio correction setting chosen for the rendering.

**Comment by chungy** [comment]
I believe this issue is now resolved as of https://github.com/chocolate-doom/chocolate-doom/commit/76d72fc48dc88fac4faa8b0b3907679eb8555778

With aspect ratio correction disabled, the output PNG files are 320×200.
